### DISCUSSION TITLE
Why is UserAgent not a public property?

### DISCUSSION THREAD

**Discussion by lahell** [PowerShell/PowerShell#16400] [Q&A]
I'm writing a module for a vendor API and decided I wanted to append ModuleName/Version to the default useragent in PowerShell. I located [UserAgent in Microsoft.PowerShell.Commands.PSUserAgent](https://github.com/PowerShell/PowerShell/blob/658837323599ab1c7a81fe66fcd43f7420e4402b/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/PSUserAgent.cs#L19) but noticed it's internal. Is there a particular reason why this property is hidden?

I found a workaround (see below), but not sure if this is a good idea. Should I just discard the idea of appending to the existing useragent and simply set it to ModuleName/Version?

```
$Module = $MyInvocation.MyCommand.ScriptBlock.Module.Name
$Version = $MyInvocation.MyCommand.ScriptBlock.Module.Version

try {
    $UserAgent = [Microsoft.PowerShell.Commands.PSUserAgent].GetProperty(
        'UserAgent',
        [System.Reflection.BindingFlags]::Static -bor
        [System.Reflection.BindingFlags]::NonPublic
    ).GetValue([Microsoft.PowerShell.Commands.PSUserAgent])
} catch {
    $UserAgent = $null
}

$UserAgent, "$Module/$Version" -join ' '
```

**Comment by iSazonov** [comment]
It is helper class for standard Web cmdlets.

You show a workaround but it is not clear how and why you use the class.

**Reply by lahell** [reply]
Sorry, should have been more clear. I thought it might be useful for the vendor if useragent contained modulename and version in addition to `Mozilla/5.0 (Windows NT; Windows NT 10.0; en-US) WindowsPowerShell/5.1.19041.1237` and after obtaining the useragent, I append modulename and version and send it with the `-UserAgent` parameter of `Invoke-RestMethod`. It doesn't matter for the functionality of my module, I was just curious if there was a reason behind the choice of making `UserAgent` non-public.

**Comment by laura-rodriguez** [comment]
Hi @iSazonov,

I have a similar scenario as @lahell, where I want to append my own userAgent to the default one. What's the recommended way considering that the UserAgent property of [Microsoft.PowerShell.Commands.PSUserAgent] is internal?

**Reply by iSazonov** [reply]
If we follow strictly [the purpose of this property](https://stackoverflow.com/questions/2601372/what-is-the-standard-format-for-a-browsers-user-agent-string), its value must unambiguously correspond to the implementation of the agent. If we say about PowerShell cmdlets, the agent is implemented in PowerShell and no one should change the property.
If you implement your agent (new binary cmdlet) in your module instead of Invoke-WebRequest, you are free to specify your own value for this property.

**Reply by laura-rodriguez** [reply]
Thanks for your response @iSazonov!

I totally understand the intent of not allowing them to edit the default user-agent string, but why not let them read it? If devs could read the default user agent, they could append any additional info they need. That's at least my use case, I want to append data to track usage (for example, the version of my module used by users) without losing the default user agent info. When I pass my user-agent to the Invoke-WebRequest command (`Invoke-WebRequest -UserAgent "myCustomModule/1.0.2"`) my user-agent ends up overriding the default value.

Am I approaching this right? What would you recommend for my use case?

**Reply by iSazonov** [reply]
PSUserAgent class is open source and you can re-implement it in your module as you need.
